### PR TITLE
[SPARK-44940][SQL] Improve performance of JSON parsing when "spark.sql.json.enablePartialResults" is enabled

### PR DESCRIPTION
**Pull request by sadikovi**
### What changes were proposed in this pull request?

The PR improves JSON parsing when `spark.sql.json.enablePartialResults` is enabled: 
- Fixes the issue when using nested arrays `ClassCastException: org.apache.spark.sql.catalyst.util.GenericArrayData cannot be cast to org.apache.spark.sql.catalyst.InternalRow`
- Improves parsing of the nested struct fields, e.g. `{"a1": "AAA", "a2": [{"f1": "", "f2": ""}], "a3": "id1", "a4": "XXX"}` used to be parsed as `|AAA|NULL    |NULL|NULL|` and now is parsed as `|AAA|[{NULL, }]|id1|XXX|`.
- Improves performance of nested JSON parsing. The initial implementation would throw too many exceptions when multiple nested fields failed to parse. When the config is disabled, it is not a problem because the entire record is marked as NULL.

The internal benchmarks show the performance improvement from slowdown of over 160% to an improvement of 7-8% compared to the master branch when the flag is enabled. I will create a follow-up ticket to add a benchmark for this regression.

### Why are the changes needed?

Fixes some corner cases in JSON parsing and improves performance when `spark.sql.json.enablePartialResults` is enabled.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

I added tests to verify nested structs, maps, and arrays can be parsed without affecting the subsequent fields in the JSON. I also updated the existing tests when `spark.sql.json.enablePartialResults` is enabled because we parse more data now.

I added a benchmark to check performance.

Before the change (master, a45a3a3d60cb97b107a177ad16bfe36372bc3e9b):
```
[info] OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~18.04-b10 on Linux 5.4.0-1045-aws
[info] Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz
[info] Partial JSON results:                     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] parse invalid JSON                                 9537           9820         452          0.0      953651.6       1.0X
```

After the change (this PR): 
```
OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~18.04-b10 on Linux 5.4.0-1045-aws
Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz
Partial JSON results:                     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
parse invalid JSON                                 3100           3106           6          0.0      309967.6       1.0X
```

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala [hunk 1]
```diff
@@ -65,3 +93,25 @@ case class StringAsDataTypeException(
...
+    fieldValue: String,
+    jsonType: JsonToken,
+    dataType: DataType) extends RuntimeException() {
+  override def getStackTrace(): Array[StackTraceElement] = new Array[StackTraceElement](0)
+  override def fillInStackTrace(): Throwable = this
+}
+
+/**
+ * No-stacktrace equivalent of `QueryExecutionErrors.emptyJsonFieldValueError`.
+ * Used for code control flow in the parser without overhead of creating a full exception.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala:112]: So we still use exceptions in the control flow, but use a trick to avoid the stacktrace overhead?

Can we avoid using stacktrace in the control flow? or at lease not in the critical code path. e.g. `fieldConverters` should return null directly if the input is empty string and `enablePartialResults` is true.

Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala:112]: That is correct, we still use exceptions to report any issues in parsing.

Throwing exceptions allows us to propagate errors across several functions. Without this, we would have to include checks in the code for every result. Also, we would have to capture the exception itself to be reported to the user. If we decided to go with returning some class to indicate failed parsing, then we would essentially be implementing exceptions. Exceptions are not that much slower, most of the time is spent on building the stacktrace which we only need when the very first exception is thrown.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 2]
```diff
@@ -420,17 +420,17 @@ class JacksonParser(
...
       dataType match {
         case FloatType | DoubleType | TimestampType | DateType =>
-          throw QueryExecutionErrors.emptyJsonFieldValueError(dataType)
+          throw EmptyJsonFieldValueException(dataType)
         case _ => null
       }
 
     case VALUE_STRING if parser.getTextLength < 1 =>
-      throw QueryExecutionErrors.emptyJsonFieldValueError(dataType)
+      throw EmptyJsonFieldValueException(dataType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:428]: the control flow here is bit convoluted, when/where is `EmptyJsonFieldValueException` being handled first?

**Review by MaxGekk — COMMENTED** [review]
> I will create a follow-up ticket to add a benchmark for this regression.

It would be better to add such benchmarks firstly, and then show perf improvements in the PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala (hunk 2 above)
Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:428]: Depending on the JSON object, `EmptyJsonFieldValueException` will be intercepted at convertObject, convertMap, or convertArray (because every JSON value will be treated as either of those three although I am not quite sure about from_json functions)and be wrapped into PartialResultException. This exception will eventually be converted into BadRecordException in `JacksonParser.parse` method and processed in `FailureSafeParser` according to the parsing mode.

Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:428]: I hope this answered your question.

**Comment by sadikovi** [comment]
I could only confirm this regression and the fix in the internal benchmarks, it could take quite some time to port the benchmark + data generation to open source. Is it possible to add it as a follow-up?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:428]: is it more efficient to return null directly in the field converter, to avoid exception creation and wrapping? I think the implementation should be simple, we return null only when the mode is PERMISSIVE and enablePartitionResult is true.

Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:428]: I don't know if this is more efficient because you would have to check the parsing mode before returning a result or throwing an exception. Also, in permissive mode, we still want to capture the exception when badRecordsPath is enabled or _corrupt_record column is provided.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:428]: oh this is a good point, we should still throw exceptions to not break other features.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 3]
```diff
@@ -454,7 +454,20 @@ class JacksonParser(
       schema.getFieldIndex(parser.getCurrentName) match {
         case Some(index) =>
           try {
-            row.update(index, fieldConverters(index).apply(parser))
+            val value = try {
+              fieldConverters(index).apply(parser)
+            } catch {
+              case PartialResultException(partialResult, e) if enablePartialResults =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:460]: why do we swallow the exception earlier? what happens if we leave the exception to the outer `catch` block?

Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:460]: Because it is more convenient to do it this way. Yes, we could leave it in the outer catch block but then I would have to copy 

```scala
row.update(index, value)
skipRow = structFilters.skipRow(row, index)
bitmask(index) = false
```
for each case of PartialResultException.

This is just a convenience thing to not duplicate the code.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:460]: Can we have a common trait of `PartialResultException`, `PartialArrayDataResultException` and `PartialMapDataResultException`, so that we don't need to duplicate code?

Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:460]: This is a good idea, let me make the change.

Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: I have updated the code but it is now more difficult to read in my opinion.

**Comment by sadikovi** [comment]
@MaxGekk @cloud-fan Can you review again? 
I have added the benchmark that shows performance almost 3x performance improvement.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 4]
```diff
@@ -3450,6 +3450,174 @@ abstract class JsonSuite
     }
   }
 
+  test("SPARK-44940: fully parse the record except f1 if partial results are enabled") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3453]: just to confirm: this PR is only for perf and we add more tests because the original PR that added partial result handling did not add enough tests?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 5]
```diff
@@ -454,11 +454,17 @@ class JacksonParser(
       schema.getFieldIndex(parser.getCurrentName) match {
         case Some(index) =>
           try {
-            row.update(index, fieldConverters(index).apply(parser))
+            val value = fieldConverters(index).apply(parser)
+            row.update(index, value)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:458]: unnecessary change now.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala (hunk 4 above)
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3453]: Not really, the PR also fixes a ClassCastException error, which should have been fixed when I added the original code for partial results parsing.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 6]
```diff
@@ -454,11 +454,17 @@ class JacksonParser(
...
             skipRow = structFilters.skipRow(row, index)
             bitmask(index) = false
           } catch {
             case e: SparkUpgradeException => throw e
+            case err: PartialValueException if enablePartialResults =>
+              badRecordException = badRecordException.orElse(Some(err.cause))
+              row.update(index, err.partialResult)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:465]: just for my education: `err.partialResult` will be null if it's a simple column, or an array/struct/map with partial reuslt if it's array/struct/map type?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala (hunk 5 above)
Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:458]: Yes, I will update.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala (hunk 6 above)
Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:464]: Actually, not. When we encounter PartialValueException/`err.partialResult`, it is already a complex value, e.g. struct/map/array. If there was an error parsing a primitive value, it would be caught in NonFatal clause and then rethrown as PartialValueException.

**Review by cloud-fan — APPROVED** [review]

**Comment by sadikovi** [comment]
cc @HyukjinKwon for review as well.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by sadikovi** [comment]
Shall we merge this or do you have any concerns or questions? I will be more than happy to answer them or follow up on the suggestions.

We may also need to backport to Spark 3.5/3.4 where this change was originally introduced, I can open a separate PR for the backport.

**Comment by HyukjinKwon** [comment]
Merged to master. It has some conflicts in branch-3.5 and 3.4.

**Comment by sadikovi** [comment]
I have opened backport PRs (linked in this PR).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 7]
```diff
@@ -589,12 +605,25 @@ class JacksonParser(
...
-          cause)
+          convertCauseForPartialResult(cause))
       case PartialResultArrayException(rows, cause) =>
         throw BadRecordException(
           record = () => recordLiteral(record),
           partialResults = () => rows,
           cause)
+      // These exceptions should never be thrown outside of JacksonParser.
+      // They are used for the control flow in the parser. We add them here for completeness
+      // since they also indicate a bad record.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:616]: Thank you for adding this.

File: sql/core/benchmarks/JsonBenchmark-results.txt [hunk 8]
```diff
@@ -3,121 +3,125 @@ Benchmark for performance of JSON parsing
...
+Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz
 Json files in the per-line mode:          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-Text read                                          1096           1105          12          4.6         219.1       1.0X
-Schema inferring                                   3818           3830          16          1.3         763.6       0.3X
-Parsing without charset                            4107           4137          32          1.2         821.4       0.3X
-Parsing with UTF-8                                 5717           5763          41          0.9        1143.3       0.2X
+Text read                                           553            579          32          9.0         110.6       1.0X
+Schema inferring                                   2195           2196           2          2.3         439.0       0.3X
+Parsing without charset                            4272           4274           3          1.2         854.3       0.1X
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/JsonBenchmark-results.txt:81]: Given the ration between `Test read` and `Parsing without charset `, is there a chance of regression in this PR?

**Review by dongjoon-hyun — COMMENTED** [review]
+1, late LGTM. Thank you, @sadikovi and all.
Given the previous exception case, I also support the backporting decision.

I changed the issue type of SPARK-44940 from `Improvement` to `Bug` to be clear.

File: sql/core/benchmarks/JsonBenchmark-results.txt (hunk 8 above)
Comment (sadikovi) [reply sql/core/benchmarks/JsonBenchmark-results.txt:81]: The overall ratio decreased because read text benchmark was 2x faster than before. Let me double check this. Now that you pointed it out, I am curious why the benchmark itself did not improve on a faster CPU when other cases did. If I confirm it is a regression, I will follow up on this in a separate PR.

**Comment by sadikovi** [comment]
Great, thanks. I will follow up on your comment regarding the possible regression in the parsing benchmark.

**Comment by sadikovi** [comment]
@dongjoon-hyun I reran the JSON benchmark and it seems like the previous results that I published were noisy. I confirmed there is no apparent regression in the patch.

I ran only `Json files in the per-line mode` benchmark for 10 iterations. Results:

Without the patch (latest master https://github.com/apache/spark/commit/eb0b09f0f2b518915421365a61d1f3d7d58b4404 with the patch reverted):
```
[info] OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~18.04-b10 on Linux 5.4.0-1045-aws
[info] Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz
[info] Json files in the per-line mode:          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Text read                                           463            476          15         10.8          92.6       1.0X
[info] Schema inferring                                   2126           2166          48          2.4         425.1       0.2X
[info] Parsing without charset                            3195           3201           4          1.6         638.9       0.1X
[info] Parsing with UTF-8                                 4129           4140           8          1.2         825.8       0.1X
```

With the patch (latest master):
```
[info] OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~18.04-b10 on Linux 5.4.0-1045-aws
[info] Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz
[info] Json files in the per-line mode:          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Text read                                           459            467           7         10.9          91.8       1.0X
[info] Schema inferring                                   2159           2198          45          2.3         431.7       0.2X
[info] Parsing without charset                            3106           3119          12          1.6         621.2       0.1X
[info] Parsing with UTF-8                                 4071           4090          10          1.2         814.2       0.1X
```

It seems the results are approximately the same as before. However, the benchmark results tend to fluctuate quite a lot. For example, when I reran the same benchmark without any code changes: 


The second run with the patch:
```
[info] OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~18.04-b10 on Linux 5.4.0-1045-aws
[info] Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz
[info] Json files in the per-line mode:          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Text read                                           458            469           9         10.9          91.5       1.0X
[info] Schema inferring                                   2147           2184          48          2.3         429.4       0.2X
[info] Parsing without charset                            3294           3308          10          1.5         658.8       0.1X
[info] Parsing with UTF-8                                 4437           4444           8          1.1         887.4       0.1X
```

I think it is fine and it is just noise in the benchmark, no apparent regression because of the patch.

**Comment by dongjoon-hyun** [comment]
Thank you for confirming that, @sadikovi . I agree with you.
